File: src/chunking/lines.ts
export interface ChunkBoundary {
  start: number;
  end: number;
}

export interface LineGroup extends ChunkBoundary {
  text: string;
}

export function mergeAdjacentChunks(
  chunks: ChunkBoundary[],
  desiredLength: number,
): ChunkBoundary[] {
  if (chunks.length === 0) {
    return [];
  }

  const first = chunks[0];
  if (!first) {
    return [];
  }

  const merged: ChunkBoundary[] = [];
  let currentStart = first.start;
  let currentEnd = first.end;
  let currentLength = currentEnd - currentStart;

  for (const group of chunks.slice(1)) {
    const start = group.start;
    const end = group.end;
    const length = end - start;

    if (currentLength + length > desiredLength) {
      merged.push({ start: currentStart, end: currentEnd });
      currentStart = start;
      currentEnd = end;
      currentLength = length;
      continue;
    }

    currentEnd = end;
    currentLength += length;
  }

  merged.push({ start: currentStart, end: currentEnd });
  return merged;
}

export function splitLinesKeepEnds(source: string): LineGroup[] {
  const groups: LineGroup[] = [];
  const re = /.*(?:\r\n|\n|\r|$)/g;
  let match = re.exec(source);
  while (match !== null) {
    const text = match[0] ?? "";
    if (text.length === 0) {
      break;
    }
    const start = match.index;
    const end = start + text.length;
    groups.push({ start, end, text });
    if (re.lastIndex >= source.length) {
      break;
    }
    match = re.exec(source);
  }
  return groups;
}

export function chunkLines(source: string, desiredLength: number): ChunkBoundary[] {
  if (!source.trim()) {
    return [];
  }

  const lineGroups = splitLinesKeepEnds(source).map(({ start, end }) => ({
    start,
    end,
  }));
  return mergeAdjacentChunks(lineGroups, desiredLength);
}

File: internal/chunking/lines.go
package chunking

import "strings"

type Boundary struct {
	Start int
	End   int
}

type LineGroup struct {
	Boundary
	Text string
}

func MergeAdjacentChunks(chunks []Boundary, desiredLength int) []Boundary {
	if len(chunks) == 0 {
		return nil
	}

	merged := make([]Boundary, 0, len(chunks))
	currentStart := chunks[0].Start
	currentEnd := chunks[0].End
	currentLength := currentEnd - currentStart

	for _, group := range chunks[1:] {
		length := group.End - group.Start
		if currentLength+length > desiredLength {
			merged = append(merged, Boundary{Start: currentStart, End: currentEnd})
			currentStart = group.Start
			currentEnd = group.End
			currentLength = length
			continue
		}
		currentEnd = group.End
		currentLength += length
	}

	merged = append(merged, Boundary{Start: currentStart, End: currentEnd})
	return merged
}

func SplitLinesKeepEnds(source string) []LineGroup {
	groups := []LineGroup{}
	start := 0
	for start < len(source) {
		i := start
		for i < len(source) && source[i] != '\n' && source[i] != '\r' {
			i++
		}
		if i < len(source) {
			if source[i] == '\r' && i+1 < len(source) && source[i+1] == '\n' {
				i += 2
			} else {
				i++
			}
		}
		groups = append(groups, LineGroup{
			Boundary: Boundary{Start: start, End: i},
			Text:     source[start:i],
		})
		start = i
	}
	return groups
}

func ChunkLines(source string, desiredLength int) []Boundary {
	if strings.TrimSpace(source) == "" {
		return nil
	}

	lines := SplitLinesKeepEnds(source)
	boundaries := make([]Boundary, len(lines))
	for i, line := range lines {
		boundaries[i] = line.Boundary
	}
	return MergeAdjacentChunks(boundaries, desiredLength)
}
